### PR TITLE
Add extra_properties to hive table properties

### PR DESCRIPTION
**Pull request by Praveen2112**
## Description
This PR allows us to pass through additional properties to Hive when creating table in Trino. 
The additional properties can provided in the following format 
```
extra_properties = map_from_entries(ARRAY[ROW('extra.property.one', 'one'),ROW('extra.property.two', 'two')]
```

In the initial version, the additional properties will not be exposed in `SHOW CREATE`. 

## Additional context and related issues
Fixes https://github.com/trinodb/trino/issues/954
Extension of https://github.com/trinodb/trino/pull/9475

## Release notes

(x) Release notes are required, with the following suggested text:

```markdown
# Section
* Introduce `extra_properties` for adding arbitrary properties to hive table. ({issue}`954`)
```

### CODE REVIEW

File: docs/src/main/sphinx/connector/hive.rst [hunk 1]
```diff
@@ -1101,6 +1101,10 @@ to the connector using a :doc:`WITH </sql/create-table-as>` clause::
       ``s3a://test/name=${name}/``. Mapped from the AWS Athena table property
       `storage.location.template <https://docs.aws.amazon.com/athena/latest/ug/partition-projection-setting-up.html#partition-projection-specifying-custom-s3-storage-locations>`_
     - ``${table_location}/${partition_name}``
+    -
+  * - ``extra_properties``
+    - Additional properties added to a hive table which are outside the scope of
+      Trino. These properties will not be exposed when we run `SHOW CREATE TABLE`.
```
Comment (findinpath) [inline docs/src/main/sphinx/connector/hive.rst:1107]:
```suggestion
Trino. These properties are not exposed when we run `SHOW CREATE TABLE`.
```

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java [hunk 2]
```diff
@@ -1157,6 +1158,24 @@ else if (avroSchemaLiteral != null) {
         tableProperties.put("numFiles", "-1");
         tableProperties.put("totalSize", "-1");
 
+        // Extra properties
+        getExtraProperties(tableMetadata.getProperties())
```
Comment (findinpath) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:1162]: `tableMetadata.getProperties()` is called over `10` times in `getEmptyTableProperties()` method.
Can you pls extract it to a variable in a preparatory commit?

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java [hunk 3]
```diff
@@ -8510,6 +8511,87 @@ public void testCreateAcidTableUnsupported()
         assertQueryFails("CREATE TABLE acid_unsupported WITH (transactional = true) AS SELECT 123 x", "FileHiveMetastore does not support ACID tables");
     }
 
+    @Test
```
Comment (findinpath) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java:8514]: Can you pls add an accuracy test for a `CTAS` statement with extra properties?

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveTableProperties.java [hunk 4]
```diff
@@ -311,4 +326,29 @@ public static Optional<Boolean> isAutoPurge(Map<String, Object> tableProperties)
...
+            Map<String, String> extraProperties = (Map<String, String>) value;
+            if (extraProperties.isEmpty()) {
+                throw new TrinoException(INVALID_TABLE_PROPERTY, "Extra table properties cannot be empty");
+            }
+            if (extraProperties.containsValue(null)) {
+                throw new TrinoException(INVALID_TABLE_PROPERTY, format("Extra table property value cannot be null '%s'", extraProperties));
+            }
+            return extraProperties.entrySet().stream()
+                    .collect(toImmutableMap(
+                            entry -> entry.getKey().toLowerCase(ENGLISH),
```
Comment (findinpath) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveTableProperties.java:350]: Why do we lowercase the properties?

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java (hunk 2 above)
Comment (Praveen2112) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:1162]: But won't it reduce the code readability ? Like we might need to introduce a new variable with `tableProperties` - which might conflict with the existing variable ?

Comment (findinpath) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:1162]: I don't think that adding a new variable `tableMetadataProperties` will reduce the readability of the code.
You can however disregard this request because it is rather a nit.

**Comment by Praveen2112** [comment]
@findinpath AC

File: docs/src/main/sphinx/connector/hive.rst [hunk 5]
```diff
@@ -1101,6 +1101,11 @@ to the connector using a :doc:`WITH </sql/create-table-as>` clause::
       ``s3a://test/name=${name}/``. Mapped from the AWS Athena table property
       `storage.location.template <https://docs.aws.amazon.com/athena/latest/ug/partition-projection-setting-up.html#partition-projection-specifying-custom-s3-storage-locations>`_
     - ``${table_location}/${partition_name}``
+    -
+  * - ``extra_properties``
```
Comment (findinpath) [inline docs/src/main/sphinx/connector/hive.rst:1105]: As discussed offline, please add a paragraph in the connector docs about how the user can specify extra properties for a hive table  and also how they can be retrieved via "$properties" metadata table.
Follow-up PR though

cc @mosabua

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java [hunk 6]
```diff
@@ -8510,6 +8511,117 @@ public void testCreateAcidTableUnsupported()
...
+                        CREATE TABLE %s (
+                          c1 integer)
+                        WITH (
+                          extra_properties = MAP(),
+                          format = 'ORC'
+                        )""",
+                tableName);
+        assertQueryFails(createTableSql, ".*Extra table properties cannot be empty.*");
+    }
+
```
Comment (findinpath) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java:8604]: Add an extra testcase for the following scenario

```
trino:default> create table test4 (c1 integer) 
            -> with (
            ->      extra_properties = map(array['MiXEd'], array['value1']));

trino:default> select "mixed" from "test4$properties";
 mixed  
--------
 value1 
(1 row)

trino:default> select "MiXEd" from "test4$properties";
 MiXEd  
--------
 value1 
(1 row)
```

We showcase here the fact that the user can retrieve, independently of the case used, the value for a table property

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java [hunk 7]
```diff
@@ -8510,6 +8511,117 @@ public void testCreateAcidTableUnsupported()
...
+    @Test
+    public void testMixedCaseProperty()
```
Comment (findinpath) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java:8606]:
```suggestion
public void testCollidingNamesakeMixedCaseProperties()
```

**Review by findinpath — APPROVED** [review]

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java (hunk 6 above)
Comment (Praveen2112) [reply plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java:8595]: I'm not sure if we need to add this test case - as it shows current limitation we have today i.e all names are converted to lower case for the connector.

**Review by findinpath — APPROVED** [review]

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveTableProperties.java [hunk 8]
```diff
@@ -173,7 +177,25 @@ public HiveTableProperties(
...
+                            Map<String, String> extraProperties = (Map<String, String>) value;
+                            if (extraProperties.isEmpty()) {
+                                throw new TrinoException(INVALID_TABLE_PROPERTY, "Extra table properties cannot be empty");
+                            }
+                            if (extraProperties.containsValue(null)) {
+                                throw new TrinoException(INVALID_TABLE_PROPERTY, format("Extra table property value cannot be null '%s'", extraProperties));
+                            }
+                            return extraProperties;
+                        },
+                        value -> value));
```
[the remark this answers is not in the document]
Comment (Praveen2112) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveTableProperties.java:198]: Unfortunately we can't use `Function#identity` as we take in a generic type i.e`Map` here and return an `Object` and `Function#identity` takes in `Object` and returns an `Object`.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java [hunk 9]
```diff
@@ -3760,7 +3760,7 @@ public Optional<CatalogSchemaTableName> redirectTable(ConnectorSession session,
                 name.getCatalogName(),
                 new SchemaTableName(
                         name.getSchemaTableName().getSchemaName(),
-                        name.getSchemaTableName().getTableName() + tableNameSplit.getSuffix().orElse(""))));
+                        name.getSchemaTableName().getTableName() + tableNameSplit.suffix().orElse(""))));
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:3782]: use `tableName.getTableName()` and remove `TableNameSplitResult` class
see
https://github.com/trinodb/trino/blob/f5084beedef43eca07d4257bff71ec7217e921b6/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/hms/TrinoHiveCatalog.java#L656-L663

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java [hunk 10]
```diff
@@ -1157,6 +1158,24 @@ else if (avroSchemaLiteral != null) {
         tableProperties.put("numFiles", "-1");
         tableProperties.put("totalSize", "-1");
 
+        // Extra properties
+        getExtraProperties(tableMetadata.getProperties())
+                .ifPresent(extraProperties -> {
+                    Set<String> existingProperties = tableProperties.buildOrThrow().keySet();
+
+                    if (existingProperties.stream().anyMatch(extraProperties::containsKey)) {
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:1166]: use `Sets.intersection(existingProperties, extraProperties.keySet())`

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java [hunk 11]
```diff
@@ -1157,6 +1158,24 @@ else if (avroSchemaLiteral != null) {
...
+                        throw new TrinoException(
+                                INVALID_TABLE_PROPERTY,
+                                "Invalid keys in extra_properties: %s".formatted(
+                                        extraProperties.keySet().stream()
+                                                .filter(existingProperties::contains)
+                                                .collect(joining(" ,"))));
+                    }
+                    for (Map.Entry<String, String> extraProperty : extraProperties.entrySet()) {
+                        tableProperties.put(extraProperty.getKey(), extraProperty.getValue());
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:1175]: What about these properties set by Trino later down the road
https://github.com/trinodb/trino/blob/5dc41d458be2af705a8433058efe2f33254ce572/plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftMetastoreUtil.java#L777-L780
?

File: docs/src/main/sphinx/connector/hive.rst [hunk 12]
```diff
@@ -1101,6 +1101,10 @@ to the connector using a :doc:`WITH </sql/create-table-as>` clause::
       ``s3a://test/name=${name}/``. Mapped from the AWS Athena table property
       `storage.location.template <https://docs.aws.amazon.com/athena/latest/ug/partition-projection-setting-up.html#partition-projection-specifying-custom-s3-storage-locations>`_
     - ``${table_location}/${partition_name}``
+  * - ``extra_properties``
+    - Additional properties added to a hive table which are outside the scope of
```
Comment (mosabua) [inline docs/src/main/sphinx/connector/hive.rst:1105]: How about

Additional properties added to a Hive table. The properties are not used by Trino, and are not shown in SHOW CREATE TABLE statements.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveTableProperties.java [hunk 13]
```diff
@@ -173,7 +177,25 @@ public HiveTableProperties(
...
+                        new MapType(VARCHAR, VARCHAR, typeManager.getTypeOperators()),
+                        Map.class,
+                        null,
+                        false,
+                        value -> {
+                            Map<String, String> extraProperties = (Map<String, String>) value;
+                            if (extraProperties.isEmpty()) {
+                                throw new TrinoException(INVALID_TABLE_PROPERTY, "Extra table properties cannot be empty");
+                            }
+                            if (extraProperties.containsValue(null)) {
```
[the remark this answers is not in the document]
Comment (Praveen2112) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveTableProperties.java:193]: `map_from_entries` doesn't allow us to create a `Map` with null keys right ? Anyways it is nice to have a check there.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveTableProperties.java [hunk 14]
```diff
@@ -173,7 +177,25 @@ public HiveTableProperties(
...
+                        EXTRA_PROPERTIES,
+                        "Extra table properties",
+                        new MapType(VARCHAR, VARCHAR, typeManager.getTypeOperators()),
+                        Map.class,
+                        null,
+                        false,
+                        value -> {
+                            Map<String, String> extraProperties = (Map<String, String>) value;
+                            if (extraProperties.isEmpty()) {
+                                throw new TrinoException(INVALID_TABLE_PROPERTY, "Extra table properties cannot be empty");
```
[the remark this answers is not in the document]
Comment (Praveen2112) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveTableProperties.java:191]: To avoid an edge case - but if we consider `emptyMap` as a default value then I guess it makes sense to remove the changes. WDYT ?

[the remark this answers is not in the document]
Comment (findepi) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveTableProperties.java:191]: What kind of edge case are you concerned about?

[the remark this answers is not in the document]
Comment (findepi) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveTableProperties.java:191]: > WDYT ?

IMO empty `extra_properties` means "no extra properties", same as omitting the property.

[the remark this answers is not in the document]
Comment (Praveen2112) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveTableProperties.java:191]: How would we handle if we run `SHOW CREATE TABLE` (we don't support it as of now - but in future). 
If the property is omitted then `extra_properties` is either empty or set to `null` value - do we need this differentiation ?

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveTableProperties.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (Praveen2112) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveTableProperties.java:193]: Added additional validation - but I'm not sure on how to create map with null keys. For `map` and `map_with_entries` it was throwing error.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java [hunk 15]
```diff
@@ -1157,6 +1158,24 @@ else if (avroSchemaLiteral != null) {
         tableProperties.put("numFiles", "-1");
         tableProperties.put("totalSize", "-1");
 
+        // Extra properties
+        getExtraProperties(tableMetadata.getProperties())
+                .ifPresent(extraProperties -> {
+                    Set<String> existingProperties = tableProperties.buildOrThrow().keySet();
```
[the remark this answers is not in the document]
Comment (Praveen2112) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java]: Moved to a slightly different implementation - but we build the two different tableProperties.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java (hunk 11 above)
Comment (Praveen2112) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java]: Have exposed `STATS_PROPERTIES` in `HiveMetadata` - but not sure if it is the correct approach as `HiveMetadata` is independent of the type of metastore like thrift or glue but not we directly import from thrift package. Any workarounds or pointer would be helpful.

**Comment by Praveen2112** [comment]
@mosabua / @findepi AC

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveTableProperties.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (Praveen2112) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveTableProperties.java]: Makes sense. Removed that validation

**Comment by findinpath** [comment]
@Praveen2112 

`trino-hive` tests are failing

```
Failures: 
Error:    TestHiveConnectorTest>BaseHiveConnectorTest.testShowColumnsFromPartitions:3552->AbstractTestQueryFramework.assertQueryFails:440 
Expecting message:
  "tableName is empty"
to match regex:
  ".*Table '.*\.tpch\.\$partitions' does not exist"
but did not.
```

**Comment by Praveen2112** [comment]
@findinpath Fixed test failure. @findepi AC

File: docs/src/main/sphinx/connector/hive.rst [hunk 16]
```diff
@@ -1165,6 +1165,10 @@ to the connector using a :doc:`WITH </sql/create-table-as>` clause::
       ``s3a://test/name=${name}/``. Mapped from the AWS Athena table property
       `storage.location.template <https://docs.aws.amazon.com/athena/latest/ug/partition-projection-setting-up.html#partition-projection-specifying-custom-s3-storage-locations>`_
     - ``${table_location}/${partition_name}``
+  * - ``extra_properties``
+    - Additional properties added to a Hive table. The properties are not used by Trino,
+      and are not shown in SHOW CREATE TABLE statements.
```
Comment (findepi) [inline docs/src/main/sphinx/connector/hive.rst:1170]: currently not shown (don't promise they won't be shown in the future)

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java [hunk 17]
```diff
@@ -1170,7 +1173,29 @@ else if (avroSchemaLiteral != null) {
...
+                .addAll(baseProperties.keySet())
+                .addAll(STATS_PROPERTIES)
+                .build();
+
+        Set<String> overlappingProperty = Sets.intersection(existingProperties, extraProperties.keySet());
+
+        if (!overlappingProperty.isEmpty()) {
+            throw new TrinoException(
+                    INVALID_TABLE_PROPERTY,
+                    "Invalid keys in extra_properties: %s".formatted(String.join(" ,", overlappingProperty)));
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:1192]: `join(" ,"` or `join(", "` ?

wouldn't overlappingProperty's toString be enough? `.... in extra_properties: " + overlappingProperty`

Invalid -> Illegal

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java [hunk 18]
```diff
@@ -1170,7 +1173,29 @@ else if (avroSchemaLiteral != null) {
...
+                .orElseGet(ImmutableMap::of);
+
+        Set<String> existingProperties = ImmutableSet.<String>builder()
+                .addAll(baseProperties.keySet())
+                .addAll(STATS_PROPERTIES)
+                .build();
+
+        Set<String> overlappingProperty = Sets.intersection(existingProperties, extraProperties.keySet());
+
+        if (!overlappingProperty.isEmpty()) {
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:1189]:
```suggestion
                .build();
        Set<String> illegalExtraProperties = Sets.intersection(existingProperties, extraProperties.keySet());
        if (!illegalExtraProperties.isEmpty()) {
```

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java [hunk 19]
```diff
@@ -1170,7 +1173,29 @@ else if (avroSchemaLiteral != null) {
...
         tableProperties.putAll(partitionProjectionService.getPartitionProjectionHiveTableProperties(tableMetadata));
 
-        return tableProperties.buildOrThrow();
+        Map<String, String> baseProperties = tableProperties.buildOrThrow();
+
+        // Extra properties
+        Map<String, String> extraProperties = getExtraProperties(tableMetadata.getProperties())
+                .orElseGet(ImmutableMap::of);
+
+        Set<String> existingProperties = ImmutableSet.<String>builder()
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:1182]: This is not accurate name, since these properties aren't "existing" yet (and STATS_PROPERTIES may or may not be set)

inline to avoid having to name this

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java [hunk 20]
```diff
@@ -3804,36 +3806,6 @@ private Optional<CatalogSchemaTableName> redirectTableToHudi(ConnectorSession se
...
 
-    private static TableNameSplitResult splitTableName(String tableName)
-    {
-        int metadataMarkerIndex = tableName.lastIndexOf('$');
-        return metadataMarkerIndex <= 0 ? // marker not found or at the beginning of tableName
-                new TableNameSplitResult(tableName, Optional.empty()) :
-                new TableNameSplitResult(tableName.substring(0, metadataMarkerIndex), Optional.of(tableName.substring(metadataMarkerIndex)));
-    }
-
-    private static class TableNameSplitResult
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:3815]: "Remove TableNameSplitResult" looks unrelated. if it indeed is, let's move to separate PR.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveTableProperties.java [hunk 21]
```diff
@@ -173,7 +177,25 @@ public HiveTableProperties(
...
                         null,
-                        false));
+                        false),
+                new PropertyMetadata<>(
+                        EXTRA_PROPERTIES,
+                        "Extra table properties",
+                        new MapType(VARCHAR, VARCHAR, typeManager.getTypeOperators()),
+                        Map.class,
+                        null,
+                        true,
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveTableProperties.java:187]: `// currently not shown in SHOW CREATE TABLE`

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java [hunk 22]
```diff
@@ -8487,6 +8488,98 @@ public void testCreateAcidTableUnsupported()
...
+        assertQuery(
+                "SELECT \"extra.property.one\", \"extra.property.two\" FROM \"%s_with_select$properties\"".formatted(tableName),
+                "SELECT 'one', 'two'");
+
+        assertUpdate("DROP TABLE %s".formatted(tableName));
+        assertUpdate("DROP TABLE %s_with_select".formatted(tableName));
+    }
+
+    @Test
+    public void testShowCreateWithExtraProperties()
```
Comment (findepi) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java:8512]: merge this with `testExtraProperties` so that you CREATE and DROP table fewer times

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java [hunk 23]
```diff
@@ -8487,6 +8488,98 @@ public void testCreateAcidTableUnsupported()
...
+    public void testExtraProperties()
+    {
+        String tableName = "create_table_with_multiple_extra_properties_" + randomNameSuffix();
+        assertUpdate("CREATE TABLE %s (c1 integer) WITH (extra_properties = MAP(ARRAY['extra.property.one', 'extra.property.two'], ARRAY['one', 'two']))".formatted(tableName));
+
+        assertQuery(
+                "SELECT \"extra.property.one\", \"extra.property.two\" FROM \"%s$properties\"".formatted(tableName),
+                "SELECT 'one', 'two'");
+
+        assertUpdate("CREATE TABLE %s_with_select (c1 integer) WITH (extra_properties = MAP(ARRAY['extra.property.one', 'extra.property.two'], ARRAY['one', 'two']))".formatted(tableName));
```
Comment (findepi) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java:8501]: you can separate this `testExtraPropertiesWithCtas`

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java [hunk 24]
```diff
@@ -8487,6 +8488,98 @@ public void testCreateAcidTableUnsupported()
...
+                .isInstanceOf(QueryFailedException.class)
+                .hasMessage("Invalid keys in extra_properties: transactional");
+
+        assertThatThrownBy(() -> assertUpdate("CREATE TABLE create_table_as_select_with_extra_properties WITH (extra_properties = MAP(ARRAY['rawDataSize'], ARRAY['1'])) AS SELECT 1 as c1"))
+                .isInstanceOf(QueryFailedException.class)
+                .hasMessage("Invalid keys in extra_properties: rawDataSize");
+    }
+
+    @Test
+    public void testAddingNullProperty()
```
Comment (findepi) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java:8552]: "Extra"

`testNullExtraProperty`

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java [hunk 25]
```diff
@@ -8487,6 +8488,98 @@ public void testCreateAcidTableUnsupported()
...
+    }
+
+    @Test
+    public void testCollidingMixedCaseProperty()
+    {
+        String tableName = "create_table_with_mixed_case_extra_properties" + randomNameSuffix();
+
+        assertUpdate("CREATE TABLE %s (c1 integer) WITH (extra_properties = MAP(ARRAY['one', 'ONE'], ARRAY['one', 'ONE']))".formatted(tableName));
+        // TODO: (https://github.com/trinodb/trino/issues/17) This should run successfully
+        assertThatThrownBy(() -> query("SELECT \"one\", \"ONE\" FROM \"%s$properties\"".formatted(tableName)))
```
Comment (findepi) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java:8569]: use `SELECT *`

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java [hunk 26]
```diff
@@ -8487,6 +8488,98 @@ public void testCreateAcidTableUnsupported()
...
+                .isInstanceOf(QueryFailedException.class)
+                .hasMessageContaining("Multiple entries with same key: one=one and one=one");
+
+        assertUpdate("CREATE TABLE %s_with_select WITH (extra_properties = MAP(ARRAY['one', 'ONE'], ARRAY['one', 'ONE'])) AS SELECT 1 as c1".formatted(tableName), 1);
+        // TODO: (https://github.com/trinodb/trino/issues/17) This should run successfully
+        assertThatThrownBy(() -> query("SELECT \"one\", \"ONE\" FROM \"%s_with_select$properties\"".formatted(tableName)))
```
Comment (findepi) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java:8575]: use `SELECT *`

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java [hunk 27]
```diff
@@ -8487,6 +8488,98 @@ public void testCreateAcidTableUnsupported()
...
+    {
+        String tableName = "create_table_with_mixed_case_extra_properties" + randomNameSuffix();
+
+        assertUpdate("CREATE TABLE %s (c1 integer) WITH (extra_properties = MAP(ARRAY['one', 'ONE'], ARRAY['one', 'ONE']))".formatted(tableName));
+        // TODO: (https://github.com/trinodb/trino/issues/17) This should run successfully
+        assertThatThrownBy(() -> query("SELECT \"one\", \"ONE\" FROM \"%s$properties\"".formatted(tableName)))
+                .isInstanceOf(QueryFailedException.class)
+                .hasMessageContaining("Multiple entries with same key: one=one and one=one");
+
+        assertUpdate("CREATE TABLE %s_with_select WITH (extra_properties = MAP(ARRAY['one', 'ONE'], ARRAY['one', 'ONE'])) AS SELECT 1 as c1".formatted(tableName), 1);
```
Comment (findepi) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java:8573]: no need to test `_with_select` case in this test method

File: docs/src/main/sphinx/connector/hive.rst (hunk 16 above)
Comment (findepi) [reply docs/src/main/sphinx/connector/hive.rst:1170]: ... can be seen in $properties

**Review by findepi — APPROVED** [review]

File: docs/src/main/sphinx/connector/hive.rst (hunk 16 above)
Comment (Praveen2112) [reply docs/src/main/sphinx/connector/hive.rst]: @mosabua Can you review the document changes.

File: docs/src/main/sphinx/connector/hive.rst [hunk 28]
```diff
@@ -1165,6 +1165,11 @@ to the connector using a :doc:`WITH </sql/create-table-as>` clause::
       ``s3a://test/name=${name}/``. Mapped from the AWS Athena table property
       `storage.location.template <https://docs.aws.amazon.com/athena/latest/ug/partition-projection-setting-up.html#partition-projection-specifying-custom-s3-storage-locations>`_
     - ``${table_location}/${partition_name}``
+  * - ``extra_properties``
+    - Additional properties added to a Hive table. The properties are not used by Trino,
+      and are currently not shown in SHOW CREATE TABLE statements. However, these properties
```
Comment (mosabua) [inline docs/src/main/sphinx/connector/hive.rst:1170]: How about 

```
Additional properties added to a Hive table. The properties are not used by Trino, and are available in the ``$properties`` metadata table. The properties are not included in the output of ``SHOW CREATE TABLE`` statements.
```

Comment (mosabua) [reply docs/src/main/sphinx/connector/hive.rst:1170]: Also note.. I used the term metadata table .. like in the Iceberg connector doc .. since I believe thats the better terms.. system table is more like from the system catalog .. with Trino as the system. Does that make sense? Also .. we will really have to add a separate section about that metadata table after this PR is merged. And maybe others

**Review by mosabua — CHANGES_REQUESTED** [review]

File: docs/src/main/sphinx/connector/hive.rst (hunk 28 above)
Comment (Praveen2112) [reply docs/src/main/sphinx/connector/hive.rst:1170]: Should we mention `The properties are not included in the output of` it could changed in the future ?

Comment (Praveen2112) [reply docs/src/main/sphinx/connector/hive.rst]: @mosabua AC

**Review by mosabua — APPROVED** [review]
Docs look good now.
